_posts/2020-11-11--kinesis-streams-part-2.md

```diff
@@ -0,0 +1,222 @@
+
+Let’s take a closer look at the shared throughput model, also known as the standard iterator.
+
+**TODO**: own version of this
+
+
+Though the getRecords API calls r hidden behind the magical ESM, there are some important [limits](https://docs.aws.amazon.com/streams/latest/dev/service-sizes-and-limits.html) to be aware of. You can have at most **5 `getRecords` calls** within one second and each call can return at most **10 000 records** or **10 MB** from your stream. It's worth noting that the getRecords call can actually return 5 times more data than the **2MB** shard limit would permit, but as a result, all the subsequent calls within the next 5 seconds will fail.
+
+All these means, that while technically there’s not limit to how many lambda functions you can attach to a single stream while using the standard iterator, in reality the 2MB throughput limit means that you might start getting behind your stream if u have more than 2 consumers, and the 5 `getRecords` call limit means, that you can have at most 5 consumers before they start to get throttled. In both cases you may start seeing _ReadProvisionedThroughputExceeded_ in your CW metrics. Though event source mapping will take care of retries behind the scenes  when this happens, this will slow down the stream processing even further and will be reflected in the _IteratorAge_ metrics.
+
+So, one way or another, if you need to have more than 2-3 applications reading from your stream, enhanced fan-out is the way to go. This will give you up to 20 lambdas with a dedicated throughput. Moreover, you can use both enhanced fan-out and shared throughput for the same stream, which adds another 2-3 lambdas that share the bandwidth. Of course, all the consumers don’t have to be lambdas, you can also mix and match different types of consumers.
+
+
+## Lambda concurrency limit
+
+Word of caution:
+One last, but by far not the least thing to note about lambda, and this one is once again about the limits (because let’s face it, there r always limits 🙂). Concurrent lambda executions is a [limited resource](https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-limits.html). By default, you can only have up to 1 000 concurrent lambda executions in the same region. And though it’s a soft limit that you can increase, there still is going to be a limit. And once that limit is reached, all the new lambda executions in the same account will be throttled.
+
+In case of lambda function as a stream consumer, there might be up to 10 times as many concurrent lambda executions as you have shards in the stream with just one consumer application. Imagine having a stream with 100 shards and a parallelisation factor of 10, and you will have 1 000 lambda executions polling your stream at all times. This means that some of your business critical lambda functions might stop working because your stream consumer takes up all of the lambda "budget" for the account/region.
+
+
+
+# Conclusions
+
```

Comment:
Kinesis now supports retention time up to a year. It was a new release (https://aws.amazon.com/about-aws/whats-new/2020/11/amazon-kinesis-data-streams-enables-data-stream-retention/) Not sure if it fits into this blog, but might be an idea to mention it to get the latest and greatest info in this.

Reply:
yeah, great point, was thinking of that myself. Will probably edit my first post. The other problem is that re:Invnt is coming and many of the things may be outdated in a month 🙈